pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -846,61 +847,155 @@ func (m *kubeGenericRuntimeManager) purgeInitContainers(ctx context.Context, pod
 
+	podHasInitialized := false
 	// If any of the main containers have status and are Running, then all init containers must
 	// have been executed at some point in the past.  However, they could have been removed
 	// from the container runtime now, and if we proceed, it would appear as if they
-	// never ran and will re-execute improperly.
+	// never ran and will re-execute improperly except for the sidecar containers.
 	for i := range pod.Spec.Containers {
 		container := &pod.Spec.Containers[i]
 		status := podStatus.FindContainerStatusByName(container.Name)
 		if status != nil && status.State == kubecontainer.ContainerStateRunning {
-			return nil, nil, true
+			podHasInitialized = true
+			break
 		}
 	}
 
-	// If there are failed containers, return the status of the last failed one.
-	for i := len(pod.Spec.InitContainers) - 1; i >= 0; i-- {
+	previousContainerInitialized := true
+	restartOnFailure := shouldRestartOnFailure(pod)
+	for i := range pod.Spec.InitContainers {
 		container := &pod.Spec.InitContainers[i]
-		status := podStatus.FindContainerStatusByName(container.Name)
-		if status != nil && isInitContainerFailed(status) {
-			return status, container, false
+		if podHasInitialized && !types.IsSidecarContainer(container) {
+			// after initialization, only sidecar containers need to be kept running
+			continue
 		}
-	}
 
-	// There are no failed containers now.
-	for i := len(pod.Spec.InitContainers) - 1; i >= 0; i-- {
-		container := &pod.Spec.InitContainers[i]
 		status := podStatus.FindContainerStatusByName(container.Name)
+		klog.V(4).InfoS("Computing init container action", "pod", klog.KObj(pod), "container", container.Name, "status", status)
+		// If the container is not found, it means it has not been started yet.
 		if status == nil {
-			continue
+			if podHasInitialized || previousContainerInitialized {
+				// For the new init container,
+				// - if the pod has been initialized before, we should restart the
+				// sidecar container here as it should restart always.
+				// - if the previous init container has been initialized, it is time to
+				// start the next one.
+				changes.InitContainersToStart = append(changes.InitContainersToStart, i)
+			}
+			// There is an init container that has not been started yet.
+			// We can stop here.
+			return false
 		}
 
-		// container is still running, return not done.
-		if status.State == kubecontainer.ContainerStateRunning {
-			return nil, nil, false
-		}
+		previousContainerInitialized = false
+		switch status.State {
+		case kubecontainer.ContainerStateCreated:
+			switch {
+			case types.IsSidecarContainer(container):
+			default: // init container
+				return false
+			}
+
+		case kubecontainer.ContainerStateRunning:
+			switch {
+			case types.IsSidecarContainer(container):
+				if container.StartupProbe != nil {
+					startup, found := m.startupManager.Get(status.ID)
+					if !found {
+						// If the startup probe has not been run, wait for it.
+						continue
```

Comment:
I might be misunderstanding, but how does `continue` result in us waiting for the probe status? Won't continue go to the next init container since this is part of the `for i := range pod.Spec.InitContainers` loop?